### ISSUE TITLE
Warning embedding policy exceeds 10% of system memory + low confidence

### ISSUE THREAD

**Issue by huberrom** [RasaHQ/rasa_core#1824]
**Rasa Core version**: 0.13.2

**Issue**: I have the following stacktrace when training with my embedding policy :

```
2019-03-13 11:17:42 DEBUG    rasa_core.training.generator  - Number of augmentation rounds is 0
2019-03-13 11:17:42 DEBUG    rasa_core.training.generator  - Starting data generation round 0 ... (with 1 trackers)
Processed Story Blocks: 100%|██████████| 135/135 [00:00<00:00, 727.23it/s, # trackers=1]
2019-03-13 11:17:42 DEBUG    rasa_core.training.generator  - Finished phase (97 training samples found).
2019-03-13 11:17:42 DEBUG    rasa_core.training.generator  - Data generation rounds finished.
2019-03-13 11:17:42 DEBUG    rasa_core.training.generator  - Found 0 unused checkpoints
2019-03-13 11:17:42 DEBUG    rasa_core.training.generator  - Found 97 training trackers.
2019-03-13 11:17:42 DEBUG    rasa_core.agent  - Agent trainer got kwargs: {}
2019-03-13 11:17:42 DEBUG    rasa_core.policies.embedding_policy  - Started training embedding policy.
2019-03-13 11:17:42 DEBUG    rasa_core.featurizers  - Creating states and action examples from collected trackers (by FullDialogueTrackerFeaturizer(LabelTokenizerSingleStateFeaturizer))...
Processed trackers: 100%|██████████| 97/97 [00:00<00:00, 23073.07it/s]
2019-03-13 11:17:42 DEBUG    rasa_core.featurizers  - The longest dialogue has 16 actions.
2019-03-13 11:17:42 DEBUG    rasa_core.policies.embedding_policy  - Check if num_neg 20 is smaller than number of actions 62, else set num_neg to the number of actions - 1
2019-03-13 11:17:53.145619: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-03-13 11:17:53 INFO     rasa_core.policies.embedding_policy  - Accuracy is updated every 20 epochs
Epochs:   0%|          | 0/2000 [00:00<?, ?it/s]2019-03-13 11:17:59.428659: W tensorflow/core/framework/allocator.cc:122] Allocation of 39259392 exceeds 10% of system memory.
2019-03-13 11:18:00.179547: W tensorflow/core/framework/allocator.cc:122] Allocation of 39259392 exceeds 10% of system memory.
Epochs:   1%|          | 19/2000 [00:33<48:10,  1.46s/it, loss=0.410, acc=0.450]2019-03-13 11:18:28.421940: W tensorflow/core/framework/allocator.cc:122] Allocation of 39259392 exceeds 10% of system memory.
2019-03-13 11:18:28.441308: W tensorflow/core/framework/allocator.cc:122] Allocation of 39259392 exceeds 10% of system memory.
Epochs:   2%|▏         | 39/2000 [01:02<48:21,  1.48s/it, loss=0.284, acc=0.841]2019-03-13 11:18:57.501154: W tensorflow/core/framework/allocator.cc:122] Allocation of 39259392 exceeds 10% of system memory.
Epochs: 100%|██████████| 2000/2000 [34:47<00:00,  1.02s/it, loss=0.099, acc=0.992]
2019-03-13 11:52:40 INFO     rasa_core.policies.embedding_policy  - Finished training embedding policy, loss=0.099, train accuracy=0.992
2019-03-13 11:52:40 DEBUG    rasa_core.featurizers  - Creating states and action examples from collected trackers (by MaxHistoryTrackerFeaturizer(SingleStateFeaturizer))...
Processed trackers: 100%|██████████| 97/97 [00:00<00:00, 696.27it/s, # actions=378]
2019-03-13 11:52:40 DEBUG    rasa_core.featurizers  - Created 378 action examples.
Processed actions: 378it [00:00, 3328.02it/s, # examples=378]
2019-03-13 11:52:40 DEBUG    rasa_core.policies.memoization  - Memorized 378 unique examples.
2019-03-13 11:52:42 INFO     rasa_core.agent  - Persisted model to '/app/models/chatbot-masante/core'
```

With this warning : Allocation of 39259392 exceeds 10% of system memory.

My config policy looks like this : 

```
policies:
  - name: EmbeddingPolicy
    epochs: 2000
    attn_shift_range: 3
    featurizer:
      - name: FullDialogueTrackerFeaturizer
        state_featurizer:
          - name: LabelTokenizerSingleStateFeaturizer
  - name: FallbackPolicy
    fallback_action_name: 'action_default_fallback'
    nlu_threshold: 0.7
    core_threshold: 0.3
  - name: MemoizationPolicy
    max_history: 6
```

And then, when I try to use it, I have really low confidence for the wrong intent :

```
rasa_core_1  | 2019-03-13 13:03:32 DEBUG    rasa_core.processor  - Predicted next action 'action_reponse_prise_en_charge_specialite' with prob 0.13.
```

With the keras policy, I have the right intent with prob 1.0.

**Comment by akelad** [comment]
no this just means that it's exceeding the standard allocation of memory for tensorflow (which is completely fine). I can see that the accuracy for your training is only 0.92 though, with the embedding policy that should actually reach 1.0 so I think you should add some more epochs. Please try that and let me know if that fixes it

**Comment by huberrom** [comment]
I'll try with 4000 epochs, I'll come back to you once the training is done

**Comment by huberrom** [comment]
So with 4000 epochs, I still don't have 1.0 accuracy, and the problem persists :

```
2019-03-14 13:14:36 DEBUG    rasa_core.training.generator  - Number of augmentation rounds is 0
2019-03-14 13:14:36 DEBUG    rasa_core.training.generator  - Starting data generation round 0 ... (with 1 trackers)
Processed Story Blocks: 100%|██████████| 159/159 [00:00<00:00, 887.63it/s, # trackers=1]
2019-03-14 13:14:36 DEBUG    rasa_core.training.generator  - Finished phase (106 training samples found).
2019-03-14 13:14:36 DEBUG    rasa_core.training.generator  - Data generation rounds finished.
2019-03-14 13:14:36 DEBUG    rasa_core.training.generator  - Found 0 unused checkpoints
2019-03-14 13:14:36 DEBUG    rasa_core.training.generator  - Found 106 training trackers.
2019-03-14 13:14:36 DEBUG    rasa_core.agent  - Agent trainer got kwargs: {}
2019-03-14 13:14:36 DEBUG    rasa_core.policies.embedding_policy  - Started training embedding policy.
2019-03-14 13:14:36 DEBUG    rasa_core.featurizers  - Creating states and action examples from collected trackers (by FullDialogueTrackerFeaturizer(LabelTokenizerSingleStateFeaturizer))...
Processed trackers: 100%|██████████| 106/106 [00:00<00:00, 25500.21it/s]
2019-03-14 13:14:36 DEBUG    rasa_core.featurizers  - The longest dialogue has 16 actions.
2019-03-14 13:14:36 DEBUG    rasa_core.policies.embedding_policy  - Check if num_neg 20 is smaller than number of actions 62, else set num_neg to the number of actions - 1
2019-03-14 13:14:44.785727: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-03-14 13:14:45 INFO     rasa_core.policies.embedding_policy  - Accuracy is updated every 20 epochs
Epochs: 100%|██████████| 4000/4000 [1:17:44<00:00,  1.03s/it, loss=0.058, acc=0.987]
2019-03-14 14:32:30 INFO     rasa_core.policies.embedding_policy  - Finished training embedding policy, loss=0.058, train accuracy=0.987
2019-03-14 14:32:30 DEBUG    rasa_core.featurizers  - Creating states and action examples from collected trackers (by MaxHistoryTrackerFeaturizer(SingleStateFeaturizer))...
Processed trackers: 100%|██████████| 106/106 [00:00<00:00, 882.19it/s, # actions=409]
2019-03-14 14:32:30 DEBUG    rasa_core.featurizers  - Created 409 action examples.
Processed actions: 409it [00:00, 4378.14it/s, # examples=409]
2019-03-14 14:32:30 DEBUG    rasa_core.policies.memoization  - Memorized 409 unique examples.
2019-03-14 14:32:31 INFO     rasa_core.agent  - Persisted model to '/app/models/chatbot-masante/core'
```

I used this config :
```
policies:
  - name: EmbeddingPolicy
    epochs: 4000
    attn_shift_range: 2
    featurizer:
      - name: FullDialogueTrackerFeaturizer
        state_featurizer:
          - name: LabelTokenizerSingleStateFeaturizer
  - name: FallbackPolicy
    fallback_action_name: 'action_default_fallback'
    nlu_threshold: 0.7
    core_threshold: 0.3
  - name: MemoizationPolicy
    max_history: 8
```

**Comment by akelad** [comment]
what training command did you use? And are you sure your stories don't have some conflicting examples?

**Comment by huberrom** [comment]
I use the following command :
 ```
python3 -m rasa_core.train -d domain.yml -s ./data/stories/ -o models/project_name/core -c embedding_config.yml  --debug --augmentation 0
```

Hum, maybe we have, I'm looking at them right now, but could a conflicting examples impact all the story ? Because the policy seems to failed for the basic one for me

**Comment by akelad** [comment]
could you try setting the `attn_shift_range` to 5?

**Comment by huberrom** [comment]
Doesn't seems to help, I still don't have a better accuracy :

```
2019-03-15 09:42:39 DEBUG    rasa_core.training.generator  - Number of augmentation rounds is 0
2019-03-15 09:42:39 DEBUG    rasa_core.training.generator  - Starting data generation round 0 ... (with 1 trackers)
Processed Story Blocks: 100%|██████████| 159/159 [00:00<00:00, 430.61it/s, # trackers=1]
2019-03-15 09:42:39 DEBUG    rasa_core.training.generator  - Finished phase (106 training samples found).
2019-03-15 09:42:39 DEBUG    rasa_core.training.generator  - Data generation rounds finished.
2019-03-15 09:42:39 DEBUG    rasa_core.training.generator  - Found 0 unused checkpoints
2019-03-15 09:42:39 DEBUG    rasa_core.training.generator  - Found 106 training trackers.
2019-03-15 09:42:39 DEBUG    rasa_core.agent  - Agent trainer got kwargs: {}
2019-03-15 09:42:39 DEBUG    rasa_core.policies.embedding_policy  - Started training embedding policy.
2019-03-15 09:42:39 DEBUG    rasa_core.featurizers  - Creating states and action examples from collected trackers (by FullDialogueTrackerFeaturizer(LabelTokenizerSingleStateFeaturizer))...
Processed trackers: 100%|██████████| 106/106 [00:00<00:00, 14798.17it/s]
2019-03-15 09:42:39 DEBUG    rasa_core.featurizers  - The longest dialogue has 16 actions.
2019-03-15 09:42:40 DEBUG    rasa_core.policies.embedding_policy  - Check if num_neg 20 is smaller than number of actions 62, else set num_neg to the number of actions - 1
2019-03-15 09:42:51.535299: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-03-15 09:42:52 INFO     rasa_core.policies.embedding_policy  - Accuracy is updated every 20 epochs
Epochs:   0%|          | 0/4000 [00:00<?, ?it/s]2019-03-15 09:42:57.898942: W tensorflow/core/framework/allocator.cc:122] Allocation of 40473600 exceeds 10% of system memory.
2019-03-15 09:42:58.431305: W tensorflow/core/framework/allocator.cc:122] Allocation of 40473600 exceeds 10% of system memory.
Epochs:   0%|          | 19/4000 [00:31<1:25:43,  1.29s/it, loss=0.362, acc=0.318]2019-03-15 09:43:24.564885: W tensorflow/core/framework/allocator.cc:122] Allocation of 40473600 exceeds 10% of system memory.
2019-03-15 09:43:24.586737: W tensorflow/core/framework/allocator.cc:122] Allocation of 40473600 exceeds 10% of system memory.
Epochs:   1%|          | 39/4000 [00:57<1:23:35,  1.27s/it, loss=0.256, acc=0.887]2019-03-15 09:43:51.296025: W tensorflow/core/framework/allocator.cc:122] Allocation of 40473600 exceeds 10% of system memory.
Epochs: 100%|██████████| 4000/4000 [1:24:19<00:00,  1.27s/it, loss=0.098, acc=0.981]
2019-03-15 11:07:11 INFO     rasa_core.policies.embedding_policy  - Finished training embedding policy, loss=0.098, train accuracy=0.981
2019-03-15 11:07:11 DEBUG    rasa_core.featurizers  - Creating states and action examples from collected trackers (by MaxHistoryTrackerFeaturizer(SingleStateFeaturizer))...
Processed trackers: 100%|██████████| 106/106 [00:00<00:00, 385.76it/s, # actions=421]
2019-03-15 11:07:12 DEBUG    rasa_core.featurizers  - Created 421 action examples.
Processed actions: 421it [00:00, 1865.68it/s, # examples=421]
2019-03-15 11:07:12 DEBUG    rasa_core.policies.memoization  - Memorized 421 unique examples.
2019-03-15 11:07:14 INFO     rasa_core.agent  - Persisted model to '/app/models/chatbot-masante/core'
```

**Comment by akelad** [comment]
hmm, @Ghostvv could you take a look?

**Comment by Ghostvv** [comment]
for clarification, it only doesn't work during prediction? Are you sure, you load the correct model?

**Comment by huberrom** [comment]
Well it looks like it works during evaluation, so yeah I would say only during prediction.
I'm pretty sure that I'm loading the correct model too.

I tried to check my story, using interactive learning to see if my stories were the same and I noticed a few things :

-  I have an unfeaturized slot, that I didn't put in my stories (because it shouldn't influence them), but the interactive learning put them in the exported story

- This slot should always be filled. It's filled by the chat during the welcome message (we send the intention /welcome + the slot) and this slot is never reset (during the reset conversation action, we reset the conversation and reset the slot). Doing this seems to change a little bit our story to this :

```
## Generated Story 6672823872597397743
* bienvenue{"entity": "CAAS"}
    - slot{"entity": "CAAS"}
    - action_bienvenue

## Generated Story -1393220828327871500
    - slot{"entity": "CAAS"}
* demander_prise_en_charge_specialite{"specialite_medecine_douce": "reiki"}
    - slot{"specialite_medecine_douce": ["reiki"]}
    - action_reponse_prise_en_charge_specialite
    - slot{"reponse_action_server": "reponse_negative"}
    - slot{"specialite_medecine_douce": null}
    - action_lien_connaitre_liste_medecines_douces
```

As you can see, this slot (entity) is set during the "bienvenue" (french for welcome) story, and then it is set again before all my story (because we restart the conversation between stories, so the slot is set again). Could that impact the embedding model ?

- The slot specialite_medecine_douce is a list, do I need to put it in my story like this :
```
* demander_prise_en_charge_specialite{"specialite_medecine_douce": "reiki"}
    - slot{"specialite_medecine_douce": ["reiki"]}
``` 
or is the first intention + entity enough ?

- I used "None" instead of null in my stories for the slot specialite_medecine_douce
```
- slot{"specialite_medecine_douce": null}
```

That all the things that I see that could interfere with the embedding model. What do you think ?

**Comment by Ghostvv** [comment]
You don't need to put unfeaturized slot in the stories (interactive learning puts all the slots regardless)
I'm not sure about `None` vs `null`, I always use `null`
For the list slot please use the style of interactive learning

**Comment by huberrom** [comment]
I'm not confortable with the style of interactive learning for list slot. In the following examples, I have two entities detected in my sentence, but the story generated by the interactive learning looks like this 

```
* demander_prise_en_charge_specialite{"specialite_medecine_douce": "r\u00e9flexologue"}
    - slot{"specialite_medecine_douce": ["sophrologue", "r\u00e9flexologue"]}
    - action_reponse_prise_en_charge_specialite
    - slot{"reponse_action_server": "reponse_multiple"}
    - slot{"specialite_medecine_douce": null}
    - action_reponse_liste_medecines_douces
```

despite the fact that both entities are detected :

```
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 2                                                                                                                                                        Est-ce que la [sophrologie](specialite_medecine_douce:sophrologue) et la
                                                                                                                                         [réflexologie](specialite_medecine_douce:réflexologue) sont remboursées par mon contrat ?
                                                                                                                                                                                  intent: demander_prise_en_charge_specialite 0.95
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 3    slot{"specialite_medecine_douce": ["sophrologue", "r\u00e9flexologue"]}
      action_reponse_prise_en_charge_specialite 1.00
      slot{"reponse_action_server": "reponse_multiple"}
      slot{"specialite_medecine_douce": null}
      action_reponse_liste_medecines_douces 0.99
      reponse_liste_medecines_douces
```

I think the intent part of the story should look like this :

```
demander_prise_en_charge_specialite{"specialite_medecine_douce": ["sophrologie", "réflexologie"]}
```

**Comment by Ghostvv** [comment]
thanks for spotting this! There might be a bug in printing and possibly featurizing list entities. Do you mind checking it and creating a PR?
